### PR TITLE
RAS-933 Auth SQL logs print email

### PR DESCRIPTION
**Pull request by arroyoAle**
# What and why?
This PR is to stop auth from logging the user's email when a database error occurs. The tokens method was missing a try catch compared to the rest of the repo and it would have been too much effort to update the other endpoint so we could have the database error handling only in the decorator.
# How to test?
- deploy this pr to your environment
- ensure login work normally for rops and fronstage
- singout of frontstage
- run `kubectl delete statefulsets.app postgres`
- try logging into frontstage
- check the auth logs and make sure there is no database query or multi-line stacktrace

Note: the last three steps can be done before deploying as well to see the original state of the logs, but make sure to redeploy postgres and restart your deployments to reconnect to the database
# Jira
https://jira.ons.gov.uk/browse/RAS-933

### CODE REVIEW

File: ras_rm_auth_service/resources/tokens.py [hunk 1]
```diff
@@ -40,37 +43,39 @@ def post_token():
...
+                user.authorise(payload.get("password"))
+            except Unauthorized as ex:
+                bound_logger.info("User is unauthorised", description=ex.description)
+                return make_response(jsonify({"title": AUTH_TOKEN_ERROR, "detail": ex.description}), 401)
+
+            logger.info("User credentials correct")
+            return make_response("", 204)
+    except SQLAlchemyError as e:
+        logger.info(f"{e.__class__.__name__} occurred when committing to database", code=e.code)
+        return make_response(jsonify({"title": AUTH_TOKEN_ERROR, "detail": e.__class__.__name__}), 500)
```
Comment (arroyoAle) [inline ras_rm_auth_service/resources/tokens.py:78]: 500 seems the most appropriate to me in this case

**Review by SteveScorfield — APPROVED** [review]
Tested and works as expected. Compared before and after code change messages and looks correct. Approved.

**Review by LJBabbage — CHANGES_REQUESTED** [review]
A couple of things to look at. There is at least one place that a re-raised SQLAlchemyError is not caught see batch_process_endpoints, there could be more, but I didn't check all 20 uses. Unless I have missed it I also didn't see a test for this

File: ras_rm_auth_service/db_session_handlers.py [hunk 2]
```diff
@@ -18,7 +18,7 @@ def transactional_session():
     except SQLAlchemyError as e:
         # not logging exception as it may have sensitive data
         logger.error(
```
[the remark this answers is not in the document]
Comment (arroyoAle) [reply ras_rm_auth_service/db_session_handlers.py]: Updated this to info logging

File: ras_rm_auth_service/db_session_handlers.py [hunk 3]
```diff
@@ -18,7 +18,7 @@ def transactional_session():
...
-            "Error committing to database",
+            f"Rolling back database session due to {e.__class__.__name__}",
```
[the remark this answers is not in the document]
Comment (arroyoAle) [reply ras_rm_auth_service/db_session_handlers.py]: Reverted my changes and adding the action after the cause

File: ras_rm_auth_service/db_session_handlers.py [hunk 4]
```diff
@@ -18,7 +18,7 @@ def transactional_session():
...
             error_class=e.__class__.__name__,
```
[the remark this answers is not in the document]
Comment (arroyoAle) [reply ras_rm_auth_service/db_session_handlers.py:22]: Removed the e.class.name from the string, I'll leave the other one in so at lease we know what kind of SQL error is being raised

File: ras_rm_auth_service/db_session_handlers.py [hunk 5]
```diff
@@ -18,7 +18,7 @@ def transactional_session():
...
             pool_size=current_app.db.engine.pool.size(),
             connections_in_pool=current_app.db.engine.pool.checkedin(),
```
[the remark this answers is not in the document]
Comment (arroyoAle) [reply ras_rm_auth_service/db_session_handlers.py]: It was like this before, not sure it adds anything useful either and it has been around since 2018 - has been removed

File: ras_rm_auth_service/db_session_handlers.py [hunk 6]
```diff
@@ -27,6 +27,10 @@ def transactional_session():
         )
         session.rollback()
         raise
+    except Exception as e:
```
[the remark this answers is not in the document]
Comment (arroyoAle) [reply ras_rm_auth_service/db_session_handlers.py]: It can't, has been removed

File: ras_rm_auth_service/db_session_handlers.py [hunk 7]
```diff
@@ -27,6 +27,10 @@ def transactional_session():
...
+        logger.error("Unknown error raised when committing to database", error_class=e.__class__.__name__)
```
[the remark this answers is not in the document]
Comment (arroyoAle) [reply ras_rm_auth_service/db_session_handlers.py]: Has been removed

File: ras_rm_auth_service/db_session_handlers.py [hunk 8]
```diff
@@ -48,5 +52,8 @@ def non_transactional_session():
             current_overflow=current_app.db.engine.pool.overflow(),
         )
         raise
+    except Exception as e:
```
[the remark this answers is not in the document]
Comment (arroyoAle) [reply ras_rm_auth_service/db_session_handlers.py]: Has been removed

File: ras_rm_auth_service/resources/tokens.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arroyoAle) [reply ras_rm_auth_service/resources/tokens.py:77]: Yeah I agree, but this endpoint seems to need changes on top of this, for this error maybe a 503 would be appropriate?

File: ras_rm_auth_service/db_session_handlers.py [hunk 9]
```diff
@@ -39,13 +35,9 @@ def non_transactional_session():
         yield session
     except SQLAlchemyError as e:
         # not logging exception as it may have sensitive data
-        logger.error(
+        logger.info(
```
Comment (LJBabbage) [inline ras_rm_auth_service/db_session_handlers.py:38]: I can't find any use of non_transactional_session at all, so it can be deleted

File: ras_rm_auth_service/batch_process_endpoints.py [hunk 10]
```diff
@@ -83,6 +85,8 @@ def get_users_eligible_for_first_notification():
     except NoResultFound:
         logger.info("No existing user eligible for first due deletion notification")
         return {}, 404
+    except SQLAlchemyError as e:
```
Comment (LJBabbage) [inline ras_rm_auth_service/batch_process_endpoints.py:88]: A bit of duplication that could be rationalised, but nothing major

File: ras_rm_auth_service/resources/tokens.py [hunk 11]
```diff
@@ -40,37 +43,38 @@ def post_token():
...
+            try:
+                user.authorise(payload.get("password"))
+            except Unauthorized as ex:
+                bound_logger.info("User is unauthorised", description=ex.description)
+                return make_response(jsonify({"title": AUTH_TOKEN_ERROR, "detail": ex.description}), 401)
+
+            logger.info("User credentials correct")
+            return make_response("", 204)
+    except SQLAlchemyError as e:
+        return make_response(jsonify({"title": AUTH_TOKEN_ERROR, "detail": e.__class__.__name__}), 500)
```
Comment (LJBabbage) [inline ras_rm_auth_service/resources/tokens.py:77]: Don't worry, 500 is fine, it was more a general moan and something that another story could/should be raised. Just not a fan of generic 500s

File: test/resources/test_tokens.py [hunk 12]
```diff
@@ -281,3 +284,25 @@ def test_obfuscate_email(self):
...
+
+        session_mock.side_effect = SQLAlchemyError()
+
+        # When
+        form_data = {"username": "testuser@email.com", "password": "password"}
+        response = self.client.post("/api/v1/tokens/", data=form_data, headers=self.headers)
+
+        # Then
+        self.assertEqual(response.status_code, 500)
+        self.assertEqual(response.get_json(), {"detail": "SQLAlchemyError", "title": "Auth service tokens error"})
```
Comment (LJBabbage) [inline test/resources/test_tokens.py:308]: Didn't you put in constants in the file, you should use them in asserts if you did

File: test/resources/test_tokens.py [hunk 13]
```diff
@@ -281,3 +284,25 @@ def test_obfuscate_email(self):
 
         for scenario in test_scenarios:
             self.assertEqual(obfuscate_email(scenario[0]), scenario[1])
+
+    @patch("ras_rm_auth_service.resources.tokens.transactional_session")
+    def test_post_tokens_with_database_disconnect(self, session_mock):
+        """
```
Comment (LJBabbage) [inline test/resources/test_tokens.py:290]: You don't usually put in a doc string when using given, when, then. That's kind of the idea of them, if the title isn't enough you explain it as you step through the code

File: test/resources/test_account.py [hunk 14]
```diff
@@ -416,6 +416,20 @@ def test_user_account_bad_request(self):
 
         self.assertEqual(response.status_code, 405)
 
+    @patch("ras_rm_auth_service.resources.account.transactional_session")
+    def test_get_user_database_disconnect(self, session_mock):
```
Comment (LJBabbage) [inline test/resources/test_account.py:420]: The code isn't causing a database_disconnect, it's just dealing with a SQLAlchemyError, the name should reflect that.

File: test/resources/test_tokens.py [hunk 15]
```diff
@@ -281,3 +284,25 @@ def test_obfuscate_email(self):
 
         for scenario in test_scenarios:
             self.assertEqual(obfuscate_email(scenario[0]), scenario[1])
+
+    @patch("ras_rm_auth_service.resources.tokens.transactional_session")
+    def test_post_tokens_with_database_disconnect(self, session_mock):
```
Comment (LJBabbage) [inline test/resources/test_tokens.py:289]: As above

File: test/resources/test_tokens.py [hunk 16]
```diff
@@ -281,3 +284,25 @@ def test_obfuscate_email(self):
...
+        self.client.post("/api/account/create", data=form_data, headers=self.headers)
+
+        form_data = {"username": "testuser@email.com", "account_verified": "true"}
+        self.client.put("/api/account/create", data=form_data, headers=self.headers)
+
+        session_mock.side_effect = SQLAlchemyError()
+
+        # When
+        form_data = {"username": "testuser@email.com", "password": "password"}
+        response = self.client.post("/api/v1/tokens/", data=form_data, headers=self.headers)
```
Comment (LJBabbage) [inline test/resources/test_tokens.py:304]: Does it need all of this? just the error is fine. If you really want to keep this you should try to avoid reassign vars, it's technically fine here, but makes it hard to follow in the code

File: test/test_batch_process_endpoints.py [hunk 17]
```diff
@@ -468,6 +468,26 @@ def test_get_users_eligible_for_fist_notification_with_no_result(self):
         self.assertNotIn(self.user_1, users)
         self.assertNotIn(self.user_3, users)
 
+    @patch("ras_rm_auth_service.batch_process_endpoints.transactional_session")
+    def test_get_users_eligible_for_first_notification_with_database_disconnect(self, session_mock):
```
Comment (LJBabbage) [inline test/test_batch_process_endpoints.py:472]: As above with name

File: test/test_batch_process_endpoints.py [hunk 18]
```diff
@@ -468,6 +468,26 @@ def test_get_users_eligible_for_fist_notification_with_no_result(self):
...
+
+        # When:
+        _datetime_24_months_ago = datetime.utcnow() - timedelta(days=750)
+        criteria = {"last_login_date": _datetime_24_months_ago}
+        criteria_one = {"account_creation_date": _datetime_24_months_ago}
+        self.update_test_data(self.user_0, criteria)
+        self.update_test_data(self.user_2, criteria)
+        self.update_test_data(self.user_1, criteria_one)
+        self.update_test_data(self.user_3, criteria_one)
+        response = self.client.get("/api/batch/account/users/eligible-for-first-notification", headers=self.headers)
```
Comment (LJBabbage) [inline test/test_batch_process_endpoints.py:485]: I haven't dug into the code, but can you get away without all of this, we just want to raise an error

**Review by LJBabbage — CHANGES_REQUESTED** [review]
All works and looks good, some minor stuff, I didn't mark everything as it was duplication, i.e tests names. I do think you are possibly mocking to low, we should have one test that exercises the decorator all the others can be at view level. I.e whatever calls the  methods that has the decorator on should be mocked to just return SQLAlchemyError

File: test/resources/test_tokens.py (hunk 12 above)
Comment (arroyoAle) [reply test/resources/test_tokens.py]: Updated this

File: test/resources/test_tokens.py (hunk 13 above)
Comment (arroyoAle) [reply test/resources/test_tokens.py]: Removed the docstring

File: test/resources/test_account.py (hunk 14 above)
Comment (arroyoAle) [reply test/resources/test_account.py]: Updated name

File: test/resources/test_tokens.py (hunk 16 above)
Comment (arroyoAle) [reply test/resources/test_tokens.py]: Removed and still works

File: test/test_batch_process_endpoints.py (hunk 17 above)
Comment (arroyoAle) [reply test/test_batch_process_endpoints.py]: Updated name

File: test/test_batch_process_endpoints.py (hunk 18 above)
Comment (arroyoAle) [reply test/test_batch_process_endpoints.py]: Removed and still works

File: ras_rm_auth_service/db_session_handlers.py (hunk 9 above)
Comment (arroyoAle) [reply ras_rm_auth_service/db_session_handlers.py]: Removed non_transactional_session

File: ras_rm_auth_service/batch_process_endpoints.py (hunk 10 above)
Comment (arroyoAle) [reply ras_rm_auth_service/batch_process_endpoints.py:88]: Not sure which bit you mean honestly

Comment (LJBabbage) [reply ras_rm_auth_service/batch_process_endpoints.py:88]: They all return 
```return make_response(jsonify({"title": BATCH_PROCESS_ERROR, "detail": e.__class__.__name__}), 500)```

**Review by LJBabbage — APPROVED** [review]
